sinks/grpsink/grpc_stream.go

```diff
@@ -0,0 +1,273 @@
+func (gs *GRPCStreamingSpanSink) Ingest(ssfSpan *ssf.SSFSpan) error {
+	if err := protocol.ValidateTrace(ssfSpan); err != nil {
+		return err
+	}
+	// TODO(sdboyer) validation of span, e.g. time bounds like in datadog sink?
+
+	// Apply any common tags, superseding defaults passed in at sink creation
+	// in the event of overlap.
+	for k, v := range gs.commonTags {
+		ssfSpan.Tags[k] = v
+	}
+
+	// gRPC indicates that it is safe to send and receive on a stream at the
+	// same time, but it is not safe to simultaneously send. Guard against that
+	// with a mutex.
+	gs.streamMut.Lock()
+	err := gs.stream.Send(ssfSpan)
+	gs.streamMut.Unlock()
+
+	if err != nil {
+		atomic.AddUint32(&gs.errCount, 1)
+
+		// gRPC guarantees that an error returned from an RPC call will be of
+		// type status.Status. In the unexpected event that they're not, this
+		// call creates a dummy type, so there's no risk of panic.
+		serr := status.Convert(err)
+
+		state := gs.grpcConn.GetState()
+		// We also want to emit a log for the very first time a connection
+		// appears to go bad. This value is reset by the stream maintainer
+		// goroutine when the channel re-enters a good state.
+		if atomic.CompareAndSwapUint32(&gs.bad, 0, 1) {
+			gs.log.WithFields(logrus.Fields{
+				logrus.ErrorKey: err,
+				"target":        gs.target,
+				"name":          gs.name,
+				"chanstate":     state.String(),
+				"code":          serr.Code(),
+				"details":       serr.Details(),
+				"message":       serr.Message(),
+			}).Warn("Error sending trace to gRPC server")
+
+		} else if state == connectivity.Ready {
+			// We expect to see dropped spans while the connection is recovering
+			// from temporary failure. Logging every such failure would be a
+			// vector for write amplification. However, it's abnormal to see an
+			// error when the underlying connection is in a good (READY) state.
+			// (Note that may be possible for some spans to error before gRPC
+			// channel state transitions away from READY)
+			gs.log.WithFields(logrus.Fields{
+				logrus.ErrorKey: err,
+				"target":        gs.target,
+				"name":          gs.name,
+				"chanstate":     state.String(),
+				"code":          serr.Code(),
+				"details":       serr.Details(),
+				"message":       serr.Message(),
+			}).Error("gRPC chan state is READY but error was returned")
+		}
+	} else {
+		atomic.AddUint32(&gs.sentCount, 1)
+	}
+
+	return err
+}
+
+// Flush reports metrics about the number of failed and successful
+//
+// No data is sent to the target sink on Flush(); this sink operates
+// entirely over a stream on Ingest().
+func (gs *GRPCStreamingSpanSink) Flush() {
+	samples := &ssf.Samples{}
+	samples.Add(ssf.Count(
+		sinks.MetricKeyTotalSpansFlushed,
+		float32(atomic.LoadUint32(&gs.sentCount)),
+		map[string]string{"sink": gs.Name()},
+	))
+	samples.Add(ssf.Count(
+		sinks.MetricKeyTotalSpansDropped,
+		float32(atomic.LoadUint32(&gs.errCount)),
+		map[string]string{"sink": gs.Name()},
+	))
```

Comment:
you can pass varargs to `samples.Add` - probably easier to skim that way

Reply:
👍